File: backend/schemas/cohort.py
from typing import Optional
from datetime import datetime
import uuid
from schemas.common import BaseSchema

class CohortBase(BaseSchema):
    name: str
    start_date: datetime
    end_date: Optional[datetime] = None
    max_students: Optional[int] = None
    registration_open: bool = True

class CohortCreate(CohortBase):
    course_id: uuid.UUID

class CohortUpdate(BaseSchema):
    name: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    max_students: Optional[int] = None
    registration_open: Optional[bool] = None

class CohortResponse(CohortBase):
    id: uuid.UUID
    course_id: uuid.UUID
    current_enrollment_count: int = 0
    created_at: datetime
    updated_at: datetime

File: lib/schemas/cohort.ts
import { z } from "zod"

export const CohortSchema = z.object({
  name: z.string().min(1, "Name is required"),
  start_date: z.string().min(1, "Start date is required"),
  end_date: z.string().optional().nullable(),
  max_students: z.coerce.number().min(1).optional().nullable(),
  registration_open: z.boolean().default(true),
  course_id: z.string().uuid("Invalid course ID").optional(),
})

export const CohortCreateSchema = CohortSchema.extend({
  course_id: z.string().uuid("Invalid course ID"),
})

export const CohortUpdateSchema = CohortSchema.partial()

export type CohortFormValues = z.infer<typeof CohortSchema>
export type CohortCreateValues = z.infer<typeof CohortCreateSchema>
export type CohortUpdateValues = z.infer<typeof CohortUpdateSchema>
